modules/auth2.js

```diff
@@ -0,0 +1,186 @@
+const _ = require('lodash');
+const dataManager = require('./dataManager');
+const Promise = require('bluebird');
+
+module.exports = class Auth2 {
+  constructor() {
+    // configuration
+    const setup = 'https://p12-setup.icloud.com/setup/ws/1';
+    this.conf = {
+      home: 'https://www.icloud.com',
+      setup: setup,
+      login: setup + '/login',
+      validate: setup + '/validate',
+      logout: setup + '/logout',
+      sendVerificationCode: setup + '/sendVerificationCode',
+      validateVerificationCode: setup + '/validateVerificationCode',
+      listDevices: setup + '/listDevices'
+    };
+
+    // default request object
+    this.req = request.defaults({
+      headers: {
+        'host': 'setup.icloud.com',
+        'origin': this.conf.home,
+        'referer': this.conf.home,
+        'User-Agent': 'Opera/9.52 (X11; Linux i686; U; en)'
+      },
+      jar: true,
+      json: true
+    });
+
+    this.login = Promise.coroutine(this.login.bind(this));
+    this._logout = Promise.coroutine(this._logout.bind(this));
+    this._getDevice = Promise.coroutine(this._getDevice.bind(this));
+    this._authenticate = Promise.coroutine(this._authenticate.bind(this));
+    this._refreshValidate = Promise.coroutine(this._refreshValidate.bind(this));
+  }
+
+  *_getDevice(session) {
+    const data = yield this.req.get({
+      url: this.conf.listDevices,
+      qs: session.params
+    });
+    return data.devices[0];
+  }
+
+  _sendVerificationCode(session, device) {
+    return this.req.post({
+      url: this.conf.sendVerificationCode,
+      qs: session.params,
+      json: device
+    });
+  }
+
+  _validateVerificationCode(session, device, code) {
+    device.verificationCode = code;
+    device.trustBrowser = true;
+    session.params.trustBrowser = true;
+    return this.req.post({
+      url: this.conf.validateVerificationCode,
+      qs: session.params,
+      json: device,
+      resolveWithFullResponse: true
+    });
+  }
+
+  /**
+   * Queries the /validate endpoint and fetches two key values we need:
+   * 1. 'dsInfo' is a nested object which contains the 'dsid' integer.
+   * This object doesn't exist until *after* the login has taken place,
+   * the first request will complain about a X-APPLE-WEBAUTH-TOKEN cookie
+   * 
+   * @param {Object} session
+   */
+  *_refreshValidate(session) {
+    try {
+      // make the request via the session params
+      const data = yield this.req.get({
+        url: this.conf.validate,
+        qs: session.params
+      });
+
+      // capture the dsid 
+      if (data.dsInfo && (data.dsInfo.appleId === session.user.apple_id ||
+        data.dsInfo.appleIdAliases.find(alias => alias === session.user.apple_id))) {
+        session.params.dsid = data.dsInfo.dsid;
+      }
+    } catch (error) {
+      if (error.statusCode !== 421) throw new Error(error);
+    }
+
+  }
+
+  /**
+   * Log out current user with current dsid
+   */
+  *_logout() {
+    try {
+      yield this.req.get({
```

Comment:
Maybe `yield` is optional here

Reply:
We need to await for this request to be finished to be able to catch the error.

Reply:
Understood

Reply:
I added a description to all methods.